libinteract/acPSN.py

```diff
@@ -0,0 +1,637 @@
+        for the current frame.    
+        """
+        
+        # Compute the PSN for the current frame
+        psn = self._compute_psn(psn_shape = psn_shape,
+                                dist_cut = dist_cut,
+                                prox_cut = prox_cut,
+                                nb_atoms_per_res = nb_atoms_per_res,
+                                i_const = i_const)
+        
+        # If there are more than one i_mins
+        if len(i_mins) > 1:
+            # Collect the filtered PSNs
+            psns = [self._filter_psn_by_imin(psn, i_min) \
+                    for i_min in i_mins]          
+            # Return a 3D array containing a 2D PSN for each i_min
+            return np.stack(psns, axis = -1)
+        
+        # If there is only one i_min
+        else:
+            # Return the 2D PSN corresponding to the i_min provided
+            return self._filter_psn_by_imin(psn, i_mins)
+
+    
+    def _iter_psns(self,
+                   universe,
+                   psn_shape,
+                   i_mins,
+                   dist_cut,
+                   prox_cut,
+                   i_const):
+        """Create a generator to parse one frame of the trajectory
+        at a time. At each call, it will return the PSN(s) computed
+        for that frame.  
+        """
+        
+        # Get the neighboring atoms per residue
+        nb_atoms_per_res = self._get_nb_atoms_per_res(\
+                                universe = universe,
+                                selstring = self.SELECTION_STRING,
+                                dist_cut = dist_cut)
+        # for each frame
+        for fs in universe.trajectory:
+            
+            # Log the progress in analyzing frames
+            sys.stdout.write(f"\rAnalyzing frame: {fs.frame}")
+            sys.stdout.flush()
+            
+            # Yield the PSN (or PSNs in case of multiple i_mins)
+            yield self._frame2psn(\
+                    i_mins = i_mins,
+                    psn_shape = psn_shape,
+                    dist_cut = dist_cut,
+                    prox_cut = prox_cut,
+                    nb_atoms_per_res = nb_atoms_per_res,
+                    i_const = i_const)
+
+
+
+    ########################### PUBLIC API ###########################
+
+
+
+    def get_average_psn(self,
+                        universe,
+                        universe_ref,
+                        norm_facts,
+                        i_min = None,
+                        dist_cut = 4.5,
+                        prox_cut = None,
+                        p_min = None,
+                        edge_weight = "strength"):
+        """Compute the average PSN over a set of 2D PSNs.
+        """
+
+        # If an invalid value was provided for the edge weight,
+        # raise an error
+        if edge_weight not in ("strength", "persistence"):   
+            raise ValueError("edge_weight must be either " \
```

Comment:
It was not supposed to be an f-string since "edge_weight" in this string represents the name of the parameter, but I converted it into an f-string anyway now to include the (wrong) value passed as "edge_weight" in the error message :)